libazureinit/src/logging.rs

```diff
@@ -31,6 +32,100 @@ fn initialize_tracing() -> sdktrace::Tracer {
     provider.tracer("azure-kvp")
 }
 
+fn kvp_env_filter() -> Result<EnvFilter, anyhow::Error> {
+    Ok(EnvFilter::builder().parse(
+        [
+            "WARN",
+            "azure_init=INFO",
+            "libazureinit::config::success",
+            "libazureinit::http::received",
+            "libazureinit::http::success",
+            "libazureinit::ssh::authorized_keys",
+            "libazureinit::ssh::success",
+            "libazureinit::user::add",
+            "libazureinit::status::success",
+            "libazureinit::status::retrieved_vm_id",
+            "libazureinit::health::status",
+            "libazureinit::health::report",
+        ]
+        .join(","),
+    )?)
+}
+
+// Public KVP wrapper API for library consumers
+struct KvpLayer<S: Subscriber>(Filtered<EmitKVPLayer, EnvFilter, S>);
+
+/// Emit tracing data to the Hyper-V KVP.
+///
+/// # Example
+///
+/// ```no_run
+/// # use libazureinit::logging::Kvp;
+/// use tracing_subscriber::layer::SubscriberExt;
+///
+/// # #[tokio::main]
+/// # async fn main() -> anyhow::Result<()> {
+/// let mut kvp = Kvp::new("a-unique-id")?;
+/// let registry = tracing_subscriber::Registry::default().with(kvp.layer());
+///
+/// // When it's time to shut down, doing this ensures all writes are flushed
+/// kvp.halt().await?;
+/// # Ok(())
+/// # }
+/// ```
+pub struct Kvp<S: Subscriber> {
+    layer: Option<KvpLayer<S>>,
+    /// The `JoinHandle` for the background task responsible for writing
+    /// KVP data to the file. The caller can use this handle to wait for
+    /// the writer to finish.
+    pub writer: JoinHandle<std::io::Result<()>>,
```

Comment:
I don't think this should be a public field since the cancellation token is private. Anyone who awaits this writer will never complete since the type isn't clone-able and the only way to have the cancellation token used is via `halt()` which takes `self`, consuming the instance.